Run lfmerge in separate container

rmunn opened the pull request:
This replaces #1315 with a deployment setup that uses a .json file, named `other-dependencies.json`, at the root of the repository to keep track of the lfmerge version we want to use. This can be overridden by using environment variables, e.g. for manual deployments. But most of the time when a new lfmerge version is desired, we'll bump the version number recorded in other-dependencies.json.

### Type of Change

- New feature (non-breaking change which adds functionality)

## Testing on your branch

Please describe how to test and/or verify your changes. Provide instructions so we can reproduce.  Please also provide relevant test data as necessary.  These instructions will be used for QA testing below.

- Run `make` as usual. This should pull down the version of the lfmerge Docker image defined in the `other-dependencies.json` file.
- Do a Send/Receive of a test project from Language Depot, if you have one, to verify that lfmerge still works correctly.

## Checklist

- [X] I have performed a self-review of my own code
- [X] I have reviewed the title/description of this PR which will be used as the squashed PR commit message
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added tests that prove my fix is effective or that my feature works

## qa.languageforge.org testing

Reviewers: add/replace your name below and check the box to sign-off/attest the feature works as expected on qa.languageforge.org

- [ ] Reviewer1 (YYYY-MM-DD HH:MM)
- [ ] Reviewer2 (YYYY-MM-DD HH:MM)

longrunningprocess commented on docker/Makefile:
```diff
@@ -35,7 +35,8 @@ unit-tests-ci:
 	npm install
-	docker-compose build mail app ld-api next-proxy next-app
+    # TODO: When this Makefile is moved to repo root, turn "../" below into "./"
+	LFMERGE_VERSION="$(shell jq -r .lfmerge ../other-dependencies.json)" docker-compose build mail app ld-api next-proxy next-app
```
`shell` nor `jq` can be assumed to be present on the dev's machine unless we add instructions to the README about additional local installs which I'd prefer not to do.  Let's discuss this approach further.

rmunn replied:
`shell` is a Makefile keyword, but you're right about `jq` needing to be added to the README.

megahirt replied:
I was thinking about this too.  We have several common *nix tools that we are using that have snuck in.  sed is also on the list and it's been in our scripts for a while.  

I think while it would have been nice to have a pure Windows dev on the team, both Hanna and Joseph were keen to use WSL Ubuntu and so we actually only support Linux and MacOS dev environments at this point.

Our small list of dev dependencies is now:
Docker
make
sed
bash
jq

Anything else?

longrunningprocess replied:
I had another thought that would eliminate the need for `jq`, I just don't want to keep adding install requirements...it just creates more maintenance issues.

What if we did something like this.

`docker/lfmerge/Dockerfile`:
```
FROM ghcr.io/sillsdev/lfmerge:2.0.98
```

That will eliminate the `.json` and the need for additional installs.  Grabbing the version now can use `head` and `cut`.  The `docker-compose.yml` would just need a reference to the `Dockerfile` like many of our other containers.

longrunningprocess replied:
FYI, `head` and `cut` are both available in our GHA runners: https://github.com/longrunningprocess/gha-sandbox/runs/6064688705?check_suite_focus=true

rmunn replied:
Done in commits 02b2456 and e508b93.

longrunningprocess commented on docs/RELEASE.md:
```diff
@@ -25,7 +25,7 @@ Current workflow:
 [languageforge.org](https://languageforge.org)
 
-Production deployments can be manually run with `VERSION=<some-docker-tag-or-semver> make deploy-prod`.
+Production deployments can be manually run with `VERSION=<some-docker-tag-or-semver> VERSION_LFMERGE=<some-docker-tag-or-semver> make deploy-prod`.
```
```suggestion
Production deployments can be manually run with `VERSION_APP=<some-docker-tag-or-semver> VERSION_PROXY=<some-docker-tag-or-semver> VERSION_NEXT_APP=<some-docker-tag-or-semver> VERSION_LFMERGE=<some-docker-tag-or-semver> make deploy-prod`.
```

longrunningprocess replied:
apparently I missed that in my previous work, sorry about that.

longrunningprocess commented on docker/Makefile:
```diff
@@ -35,7 +35,8 @@ unit-tests-ci:
 	npm install
-	docker-compose build mail app ld-api next-proxy next-app
+    # TODO: When this Makefile is moved to repo root, turn "../" below into "./"
+	LFMERGE_VERSION="$(shell jq -r .lfmerge ../other-dependencies.json)" docker-compose build mail app ld-api next-proxy next-app
```
did you happen to check whether doing this type of thing in the `docker-compose.yml` is possible?  I'm thinking something like `ghcr.io/sillsdev/lfmerge:${jq ...}` or 
```yml
environment:
    - LFMERGE_VERSION=`jq ...`
```

I'm not certain off the top of my head but would be better if possible.

rmunn replied:
I checked, and you can't use shell commands inside the docker-compose.yml file. I agree that it would have been simpler — it's the first thing that I tried — but no, if you want things that will change you have to pass them in as environment variables. (Or pass the docker-compose file through sed or something, but the documentation recommends env vars).

longrunningprocess replied:
I proposed a different idea that would eliminate some of these work-arounds.

longrunningprocess commented on .github/workflows/integrate-and-deploy.yml:
```diff
@@ -46,11 +49,19 @@ jobs:
+          echo ::set-output name=LFMERGE_NAMESPACE::ghcr.io/sillsdev/lfmerge
+          # LfMerge tag from parameter if provided, use .json file if exists, fallback to "latest" as default
+          TAG_LFMERGE=${{ inputs.image-tag-lfmerge }}
+          TAG_LFMERGE=${TAG_LFMERGE:-$(jq -r .lfmerge other-dependencies.json)}
```
it probably is but we need to verify `jq` is present in GHA's runners

rmunn replied:
https://github.com/actions/virtual-environments/blob/main/images/linux/Ubuntu2004-Readme.md lists `jq` version 1.6, and the 18.04 runner lists `jq` version 1.5. So we're good.

longrunningprocess replied:
see my other comments on this topic.

longrunningprocess commented on docker/docker-compose.yml:
```diff
@@ -76,6 +76,37 @@ services:
+      - LFMERGE_MONGO_DB_NAME_PREFIX=sf_
+      - LFMERGE_VERBOSE_PROGRESS=true
+    extra_hosts:
+      - "host.docker.internal:host-gateway"
```
remind me why this is required?

rmunn replied:
I don't remember right now. It had something to do with being able to access `localhost` from inside the container and have it redirected to the Mongo database mounted on a port of the host machine, but I'll have to look up why I needed that when I get back to the office.

megahirt replied:
Correct on the localhost thing.  I don't think it's for mongo though.  We have it commented out in the LF app and only use it for xdebug (which needs to access localhost)

megahirt replied:
Maybe you can remove it and it will still work?

longrunningprocess replied:
Let's try to remember why and if it is still needed, let's capture that "why" in a comment.

rmunn replied:
It's used in xdebug (see docker/app/docker-php-ext-xdebug.ini), so it's needed in the app and app-for-playwright containers. But it's not needed in the lfmerge container. Will remove.

rmunn replied:
Done in commit 6e74b41.

longrunningprocess commented on .github/workflows/integrate-and-deploy.yml:
```diff
@@ -14,6 +14,9 @@ on:
       image-tag-next-app:
         type: string
         required: true
+      image-tag-lfmerge:
```
the lfmerge version is really driven from the `.json`, I'm not sure we actually need this at all.  Was there a use-case you were imagining where we'd tweak the `staging.yml` or `production.yml` to feed this workflow a version?

rmunn replied:
I think I was picturing a use case where we'd add a workflow_dispatch trigger to `integrate-and-deploy.yml`, in which case it would be useful to have the ability to override the .json values. Say, because we've discovered that I introduced a bug in the current version of LfMerge and want to quickly roll back to the previous version.

You're right that that's not needed most of the time, but I thought having an override available might be a good idea.

longrunningprocess replied:
let's wait until that need comes up.

rmunn replied:
Done in commit d7b3e50.

longrunningprocess reviewed (changes requested):
yep, this is pretty close to what I would've done, nice work.

rmunn commented:
@longrunningprocess Finished addressing all your review comments. There's one merge conflict remaining, which shouldn't be a merge conflict (check it out in the GitHub UI and you'll see that it should be a trivial resolution), but apparently my edits have managed to confuse Git. So this might need resolving in the GitHub merge-conflict resolution UI, which will create a merge commit merging develop into this branch, the very thing I was trying to avoid. Oh well, I think it'll go away when we squash so it won't be a big deal.

longrunningprocess commented on .github/workflows/integrate-and-deploy.yml:
```diff
@@ -46,11 +46,18 @@ jobs:
           echo ::set-output name=TAG_NEXT_APP::${{ inputs.image-tag-next-app }}
+          echo ::set-output name=LFMERGE_NAMESPACE::ghcr.io/sillsdev/lfmerge
+          # LfMerge tag from parameter if provided, use .json file if exists, fallback to "latest" as default
+          TAG_LFMERGE=$(head -n 1 docker/lfmerge/Dockerfile | cut -d: -f2)
```
`head -1` also works that same if you like the shorter syntax

rmunn replied:
I do know that, but I've always preferred the `-n num` syntax. For one thing, if you want to specify a negative number then you must use the `-n` syntax (`head -n -1` displays all but the last line of the input, while `head --1` is a syntax error). And finally, I just checked and the `-#` syntax (where # is a number, of course) is actually *not documented* in the `head` manpage! It's a GNU tool, and they like their non-standard `info` tool, so I had to check the `info head` documentation, where I found the following:

> For compatibility ‘head’ also supports an obsolete option syntax ‘-[NUM][bkm][cqv]’, which is recognized only if it is specified first. NUM is a decimal number optionally followed by a size letter (‘b’, ‘k’, ‘m’) ...

"An obsolete option syntax" doesn't give me confidence that it will continue to be supported going forward. So I've gotten used to using the `-n num` syntax instead.

rmunn replied:
BTW, commit 8bd4169 fixes the comment, which still had a reference to the input parameter we removed earlier.

longrunningprocess replied:
:-D you're too funny!  You had me at the compat reasoning which I can appreciate but a `-1` to get the last line over `tail`?!  That's sheer madness 🤪

longrunningprocess commented on .github/workflows/integrate-and-deploy.yml:
```diff
@@ -46,11 +46,18 @@ jobs:
       -
         name: Establish image name
         id: image
+        env:
```
artifact from a previous age, can be removed

rmunn replied:
Done in commit 8bd4169.
